twister: enable running tests using sysbuild under QEMU

danieldegrasse opened the pull request:
While testing #48709, I realized that twister does not currently support executing QEMU when the test was built with sysbuild. This PR enables running QEMU in this scenario.

This is accomplished by running `ninja run` with the base directory being the "main" sysbuild application. This is opposed to the method that west uses to handle sysbuild, where each application is flashed in series. The primary disadvantage of this is that QEMU targets will need to configure their main application's `run` target to include all required images.

As an alternative to this method, we could attempt to generate a `run` target in the main build directory, that added each image to the emulator's command. However, this would require changes to sysbuild, as the `board.cmake` target is not currently sourced from the cmake modules that sysbuild enables

danieldegrasse committed:
scripts: west: move domains.py to a generic library folder

Since twister can make use of domains.py as well, refactor west's use
this file so that domains.py can be moved to a generic library folder.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
scripts: twister: update twister to support QEMU platforms with sysbuild

Update twister to support running QEMU platforms with sysbuild, by parsing
domains.yaml and executing the "run" target of the default application.
This will allow twister to test QEMU targets with sysbuild. It is assumed
that QEMU targets will add any external images they need in the build
phase

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
cmake: emu: ensure that QEMU_PIPE is set when using sysbuild

Ensure that QEMU_PIPE variable is set when using sysbuild. This is required
because twister will use the "main" sysbuild application as the target for
"ninja run" when testing QEMU targets, which means that the "main" build
must be aware of any QEMU_PIPE setting passed by twister at build time.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented on scripts/twister:
```diff
@@ -193,6 +193,7 @@ if not ZEPHYR_BASE:
     print(f'ZEPHYR_BASE unset, using "{ZEPHYR_BASE}"')
 
 sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/pylib/twister/"))
+sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/pylib/helpers"))
```
I think `build_helpers` would be clearer than `helpers`. Especially `helpers/domains.py` feels a bit vague.

nashif commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -852,7 +867,7 @@ def handle(self):
         logger.debug("Running %s (%s)" % (self.name, self.type_str))
         command = [self.generator_cmd]
-        command += ["-C", self.build_dir, "run"]
+        command += ["-C", domain_build, "run"]
```
when not using sysbuild, domain_build does not fit here, we should avoid having sysbuild terminology become the default, this will be confusing.

danieldegrasse replied:
Updated to use `build_dir` instead of `domain_build` for the variable name

danieldegrasse committed:
gitignore: exclude scripts/pylib/build_helpers from git ignore list

Exclude scripts/pylib/build_helpers from default gitignore behavior of
ignoring all directories whose names begin with build, as this directory
is used for build helper scripts shared by west and twister.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic reviewed (approved):
West related portions LGTM, did not review twister parts in detail

PerMac commented:
LGTM. Is there a way to test these changes?

danieldegrasse commented:
@PerMac I tested support for this by rebasing https://github.com/zephyrproject-rtos/zephyr/pull/48709 on top of it. Would you prefer that I add a test case to this PR (or enable one with sysbuild)? The reason I didn't do this initially was because there isn't a test in tree I'm aware of which would lend itself to sysbuild, and could run on QEMU (besides OpenAMP, which #48709 enables)

PerMac commented:
@danieldegrasse I see. Then I am reassured:) BTW I think it might be useful to add appropriate qemus as integration_platforms to tests from #48709 so they will get a proper testing in the CI, after this PR is in.
